File: RubyonRails/actioncable-sample/app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "room_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    message = Message.create!(content: data["message"])
    ActionCable.server.broadcast(
      "room_channel",{ message: data["message"]}
    )
  end

  private

  def render_message(message)
    ApplicationController.render(
      partial: "messages/message",
      locals: [message: message]
    )
  end
end

File: RubyonRails/actioncable-sample/app/javascript/channels/room_channel.js
import consumer from "channels/consumer"

consumer.subscriptions.create("RoomChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    document.querySelector('input[data-behavior="room_speaker"]').addEventListener('keypress', (event) => {
      if (event.key === 'Enter') {
        this.speak(event.target.value);
        event.target.value = '';
        return event.preventDefault();
      }
    });
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    const element = document.querySelector('#messages')
    element.insertAdjacentHTML('beforeend', data['message'])
  },

  speak: function(message) {
    return this.perform('speak', {message: message});
  }
});
